In [10]:
#  Ebnable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
Enter Team Member Names here (double click to edit):

- Name 1: Zackary Gill
- Name 2: Mai Loan Tran
- Name 3: Samantha Sprague
- Name 4: Lei Jiang


________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [32]:
from __future__ import print_function
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

___
Enter your answer here:

- 11314 instances in the dataset.
- Each instance represents an article that have had the words be binned.
- There are 20 classes in the dataset, with each corresponding to a different topic.
- I would expect this to work on articles of subjects that are have been around a long time and have well established terminology, but not well on the ones that are more trendy.
- The data is sparse because as seen above 88% of it is zero.


___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [35]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550
a = ds.data[idx].todense()
a_class = ds.target_names[ds.target[idx]]
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

# Enter distance comparison below for each pair of vectors:
p = 'Placeholder'
print('\n\nEuclidean Distance\n ab:', euclidean(a, b), 'ac:', euclidean(a, c), 'bc:',euclidean(b, c))
print('Cosine Distance\n ab:', cosine(a,b), 'ac:', cosine(a,c), 'bc:', cosine(b,c))
a1 = np.where(a > 0.5, 1, 0)
b1 = np.where(b > 0.5, 1, 0)
c1 = np.where(c > 0.5, 1, 0)
print('Jaccard Dissimilarity (vectors should be boolean values)\n ab:', jaccard(a1,b1), 'ac:', jaccard(a1,c1), 'bc:', jaccard(b1,c1))

print('\n\nThe most appropriate distance is...')
print("""Cosine is the best because it gives a scaled distance between the groups.
As you can see groups B and C are quite similar (which is reflected in the values),
and AB, AC are significantly larger which indicates a large dissimilarity.""")

Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles


Euclidean Distance
 ab: 1.0985184671870858 ac: 1.1891405425398234 bc: 0.9177794226661624
Cosine Distance
 ab: 0.6033714113755322 ac: 0.7070276149559529 bc: 0.4211595343347173
Jaccard Dissimilarity (vectors should be boolean values)
 ab: 1.0 ac: 1.0 bc: 0.0


The most appropriate distance is...
Cosine is the best because it gives a scaled distance between the groups.
As you can see groups B and C are quite similar (which is reflected in the values),
and AB, AC are significantly larger which indicates a large dissimilarity.


___
# Start of Live Session Assignment
___
<a id="KNN"></a>
<a href="#top">Back to Top</a>
## Using scikit-learn with KNN
Now let's use stratified cross validation with a holdout set to train a KNN model in `scikit-learn`. Use the example below to train a KNN classifier. The documentation for `KNeighborsClassifier` is here: http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html  

**Exercise for part 3**: Use the code below to test what value of `n_neighbors` works best for the given data. *Note: do NOT change the metric to be anything other than `'euclidean'`. Other distance functions are not optimized for the amount of data we are working with.* 

**Question for part 3**: What is the accuracy of the best classifier you can create for this data (by changing only the `n_neighbors` parameter)? 

In [34]:
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from IPython.html import widgets 
from sklearn.model_selection import train_test_split

#cv = StratifiedShuffleSplit(ds.target, n_iter = 1, test_size = 0.5, train_size=0.5)
#cv = StratifiedShuffleSplit(ds.target, n_iter = 1, test_size = 0.5, train_size=0.5)

# fill in the training and testing data and save as separate variables
#for trainidx, testidx in cv:
    # note that these are sparse matrices
#    X_train = ds.data[trainidx] 
#    X_test = ds.data[testidx] 
#    y_train = ds.target[trainidx]
#    y_test = ds.target[testidx]

# TO THE GRADER:
# We tried to run the exact code, but we are using Python 3.5 which caused it to not work
# and so we used this code from our project for KNN
#-------------------------------------------------------------------------------
X_train, X_test, y_train, y_test = train_test_split(ds.data, ds.target, test_size = 0.5, random_state=7)
myList = list(range(1,10))
# subsetting just the odd ones
neighbors = filter(lambda x: x % 2 != 0, myList)

K = 0
acc = -1
for k in neighbors:
    clf = KNeighborsClassifier(n_neighbors = k, weights = 'uniform', metric = 'euclidean')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    acc1 = accuracy_score(y_test, yhat)
    print("K:", k, "ACC:", acc1)
    if acc1 > acc:
        acc = acc1
        K = k

print('Accuracy of classifier with %d neighbors is: %.2f'%(K,acc))


K: 1 ACC: 0.5865299628778504
K: 3 ACC: 0.5184726887042602
K: 5 ACC: 0.4908962347534029
K: 7 ACC: 0.47427965352660423
K: 9 ACC: 0.4610217429733074
Accuracy of classifier with 1 neighbors is: 0.59


The best accuracy is 0.59 with 1 neighbor.

**Question for part 3**: With sparse data, does the use of a KDTree representation make sense? Why or Why not?

Enter your answer below:

Using a KDTree does not make sense for this data because a KDTree does not work with a large number of dimensions, which makes the density of the data sparse. In the case of this dataset we have both a large number of dimensions paired with sparse data. This makes the tree too large and computationally intensive.
___

_____
## KNN extensions - Centroids
Now lets look at a very closely related classifier to KNN, called nearest centroid. In this classifier (which is more appropriate for big data scenarios and sparse data), the training step is used to calculate the centroids for each class. These centroids are saved. Unknown attributes, at prediction time, only need to have distances calculated for each saved centroid, drastically decreasing the time required for a prediction. 

**Exercise for part 4**: Use the template code below to create a nearest centroid classifier. Test which metric has the best cross validated performance: Euclidean, Cosine, or Manhattan. In `scikit-learn` you can see the documentation for NearestCentroid here: 
- http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestCentroid.html#sklearn.neighbors.NearestCentroid

and for supported distance metrics here:
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.distance_metrics.html#sklearn.metrics.pairwise.distance_metrics

In [39]:
from sklearn.neighbors.nearest_centroid import NearestCentroid

# the parameters for the nearest centroid metric to test are:
#    l1, l2, and cosine (all are optimized)

cv = StratifiedShuffleSplit(n_splits=1, train_size=0.5, test_size = 0.5)
for train, test in cv.split(ds.data, ds.target):
    X_train = ds.data[train] 
    X_test = ds.data[test] 
    y_train = ds.target[train]
    y_test = ds.target[test]

    clf = NearestCentroid(metric='l2')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    print ('EL2.ACC:', accuracy_score(y_test, yhat))

    clf = NearestCentroid(metric='euclidean')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    print ('E.ACC:', accuracy_score(y_test, yhat))

    clf = NearestCentroid(metric='cosine')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    print ('C.ACC:', accuracy_score(y_test, yhat))

    clf = NearestCentroid(metric='manhattan')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    print ('M.ACC:', accuracy_score(y_test, yhat))

    clf = NearestCentroid(metric='l1')
    clf.fit(X_train, y_train)
    yhat = clf.predict(X_test)
    print ('ML1.ACC:', accuracy_score(y_test, yhat))
    
# fill in your code here
print('The best distance metric (one with the highest accuracy): Cosine 0.49')

C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\neighbors\nearest_centroid.py:141: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


EL2.ACC: 0.4071062400565671
E.ACC: 0.4071062400565671


C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\neighbors\nearest_centroid.py:141: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


C.ACC: 0.49248718401979846
M.ACC: 0.17765600141417712


C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\neighbors\nearest_centroid.py:141: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


ML1.ACC: 0.3128866890578045
The best distance metric is: Cosine 0.49


___
<a id="naive"></a>
<a href="#top">Back to Top</a>
## Naive Bayes Classification
Now let's look at the use of the Naive Bayes classifier. The 20 newsgroups dataset has 20 classes and about 130,000 features per instance. Recall that the Naive Bayes classifer calculates a posterior distribution for each possible class. Each posterior distribution is a multiplication of many conditional distributions: 

$${\arg \max}_{j} \left(p(class=j)\prod_{i} p(attribute=i|class=j) \right)$$

where $p(class=j)$ is the prior and $p(attribute=i|class=j)$ is the conditional probability.

**Question for part 5**: With this many classes and features, how many different conditional probabilities need to be parameterized? How many priors need to be parameterized?

Enter you answer here:
- There are 20 classes and 130107 features which means there are [20*130107 = 2,602,140] conditional probabilities.
- 20 priors need to be parameterized because there are 20 classes.

In [42]:
# Use this space for any calculations you might want to do
# Number of classes times the number of features
20*130107

2602140

___
## Naive Bayes in Scikit-learn
Scikit has several implementations of the Naive Bayes classifier: `GaussianNB`, `MultinomialNB`, and `BernoulliNB`. Look at the documentation here: http://scikit-learn.org/stable/modules/naive_bayes.html Take a look at each implementation and then answer this question: 

**Questions for part 6**: 
- If the instances contain mostly continuous attributes, would it be better to use Gaussian Naive Bayes, Multinomial Naive Bayes, or Bernoulli? And Why? 
- What if the data is sparse, does this change your answer? Why or Why not?

Enter you answer here:

- Multinomial Naive Bayes is mostly used for text classification, using counts of data. Bernoulli requires each attribute to be a binary classification (which is not continuous). Gaussian does not have either of these requirements and instead assumes the data has a normal distribution. Based on this Gaussian would be the best choice.
- No I would not change to Bernoulli because of the same reasons listed above. I would switch to Multinomial Naive Bayes because it has the smoothing parameter that works well with sparse data.

___

## Naive Bayes Comparison
For the final section of this notebook let's compare the performance of Naive Bayes for document classification. Look at the parameters for `MultinomialNB`, and `BernoulliNB` (especially `alpha` and `binarize`). 

**Exercise for part 7**: Using the example code below, change the parameters for each classifier and see how accurate you can make the classifiers on the test set. 

**Question for part 7**: Why are these implementations so fast to train? What does the `'alpha'` value control in these models (*i.e.*, how does it change the parameterizations)? 

In [48]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB

X_train, X_test, y_train, y_test = train_test_split(ds.data, ds.target, test_size = 0.5, random_state=7)

somerange1 = range(0, 5, 1)
somerange2 = range(0, 5, 1)

print("Bernoulli:")
for a in somerange1:
    for b in somerange2:
        clf_bnb = BernoulliNB(alpha=a*0.1, binarize=b*0.1)
        clf_bnb.fit(X_train, y_train)
        yhat_bnb = clf_bnb.predict(X_test)
        print("Alpha:", a*0.1, "Binarize:", b*0.1, "ACC:", accuracy_score(y_test, yhat_bnb))
print("Multinomial:")
for a in somerange1:
    clf_mnb = MultinomialNB(alpha=a*0.1)
    clf_mnb.fit(X_train, y_train)
    yhat_mnb = clf_mnb.predict(X_test)
    print("Alpha:", a*0.1, "ACC:", accuracy_score(y_test, yhat_mnb))

print("For Bernoulli the most accurate was with Alpha of 0 and Binarize of 0, ACC: 0.85")
print("For Multinomial the most accurate was with Alpha of 0, ACC: 0.86")
print('These classifiers are so fast because...', "the dataset is sparse (lots of zeros).")
print('The alpha values control...', "the additive smoothing parameter.")
#Explain why alpha = 0 is the best, based on sparse... etc.

#If we don't have very many examples of a class, we don't necessarily want to trust that class.
#we are trying to make a class not have a zero probability.

Bernoulli:


C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)


Alpha: 0.0 Binarize: 0.0 ACC: 0.8516881739437865
Alpha: 0.0 Binarize: 0.1 ACC: 0.6765069824995581


C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)


Alpha: 0.0 Binarize: 0.2 ACC: 0.3523068764362736


C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)


Alpha: 0.0 Binarize: 0.30000000000000004 ACC: 0.16475163514230157
Alpha: 0.0 Binarize: 0.4 ACC: 0.09245182959165635
Alpha: 0.1 Binarize: 0.0 ACC: 0.7947675446349656
Alpha: 0.1 Binarize: 0.1 ACC: 0.6222379353013965
Alpha: 0.1 Binarize: 0.2 ACC: 0.23864238995934242
Alpha: 0.1 Binarize: 0.30000000000000004 ACC: 0.09757822167226445
Alpha: 0.1 Binarize: 0.4 ACC: 0.0618702492487184
Alpha: 0.2 Binarize: 0.0 ACC: 0.7673678628248188
Alpha: 0.2 Binarize: 0.1 ACC: 0.5156443344528903
Alpha: 0.2 Binarize: 0.2 ACC: 0.14194802899063108
Alpha: 0.2 Binarize: 0.30000000000000004 ACC: 0.06434505921866714
Alpha: 0.2 Binarize: 0.4 ACC: 0.052324553650344705
Alpha: 0.30000000000000004 Binarize: 0.0 ACC: 0.7443874845324376
Alpha: 0.30000000000000004 Binarize: 0.1 ACC: 0.4180661127806258
Alpha: 0.30000000000000004 Binarize: 0.2 ACC: 0.09598727240586884
Alpha: 0.30000000000000004 Binarize: 0.30000000000000004 ACC: 0.05409227505745095
Alpha: 0.30000000000000004 Binarize: 0.4 ACC: 0.05091037652465971
Alpha: 0.4 B

C:\ProgramData\Anaconda3\envs\DS7331_35r\lib\site-packages\sklearn\naive_bayes.py:480: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)


Alpha: 0.0 ACC: 0.8596429202757645
Alpha: 0.1 ACC: 0.8241117199929291
Alpha: 0.2 ACC: 0.798302987449178
Alpha: 0.30000000000000004 ACC: 0.7756761534382182
Alpha: 0.4 ACC: 0.7560544458193389
For Bernoulli the most accurate was with Alpha of 0 and Binarize of 0, ACC: 0.85
For Multinomial the most accurate was with Alpha of 0, ACC: 0.86
These classifiers are so fast because... the dataset is sparse (lots of zeros).
The alpha values control... the additive smoothing parameter.


________________________________________________________________________________________________________

That's all! Please **upload your rendered notebook to blackboard** and please include **team member names** in the notebook submission.